### PR TITLE
Ackfunctor and History refactoring [9443] 

### PR DESCRIPTION
**Pull request by MiguelBarro**

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2763/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/130/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2542/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2775/)

File: src/cpp/rtps/builtin/discovery/participant/timedevent/DSClientEvent.cpp [hunk 1]
```diff
@@ -82,6 +84,6 @@ bool DSClientEvent::event()
     return restart;
 }
 
-}
+} // namespace rtps
```
Comment (rsanchez15) [inline src/cpp/rtps/builtin/discovery/participant/timedevent/DSClientEvent.cpp:87]:
```suggestion
} /* namespace rtps */
```

File: src/cpp/rtps/history/History.cpp [hunk 2]
```diff
@@ -47,6 +47,41 @@ History::~History()
     logInfo(RTPS_HISTORY, "");
 }
 
+History::const_iterator History::find_change(
+        CacheChange_t* ch)
+{
+    if (mp_mutex == nullptr)
+    {
+        logError(RTPS_HISTORY, "You need to create a RTPS Entity with this History before using it");
+        return const_iterator();
```
Comment (rsanchez15) [inline src/cpp/rtps/history/History.cpp:56]: The `const_iterator()` should be set to `changesEnd()`.

File: src/cpp/rtps/history/History.cpp [hunk 3]
```diff
@@ -182,6 +217,6 @@ void History::print_changes_seqNum2()
     std::cout << ss.str();
 }
 
-}
+} // namespace rtps
```
Comment (rsanchez15) [inline src/cpp/rtps/history/History.cpp:220]:
```suggestion
} /* namespace rtps */
```

**Review by rsanchez15 — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp [hunk 4]
```diff
@@ -792,35 +870,53 @@ bool PDPServer2::process_changes_release()
                 // everyone, time at which it is removed from history.
                 if (ch->kind == fastrtps::rtps::ChangeKind_t::ALIVE)
                 {
-                    remove_change_from_writer_history(mp_PDPWriter, mp_PDPWriterHistory, ch);
+                    if (!remove_change_from_writer_history(mp_PDPWriter, mp_PDPWriterHistory, ch))
```
Comment (rsanchez15) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp:873]: Comment on why if the change could not be removed from the writer history it should be released manually to return the change to the pool.

**Review by rsanchez15 — CHANGES_REQUESTED** [review]

File: include/fastdds/rtps/history/WriterHistory.h [hunk 5]
```diff
@@ -64,42 +68,65 @@ class WriterHistory : public History
...
+            const_iterator removal,
+            bool release = true) override;
+
+    /**
+     * Criteria to search a specific CacheChange_t on history
+     * @param chi inner change to compare
+     * @param cho outer change for comparisson
+     * @return true if chi matches cho criteria
      */
-    RTPS_DllAPI bool remove_change(CacheChange_t* a_change) override;
```
[the remark this answers is not in the document]
Comment (MiguelBarro) [reply include/fastdds/rtps/history/WriterHistory.h:74]: nope, there can be only one virtual the other must reference the virtual one.

File: include/fastdds/rtps/history/WriterHistory.h [hunk 6]
```diff
@@ -64,42 +68,65 @@ class WriterHistory : public History
...
+            = std::chrono::steady_clock::now() + std::chrono::hours(24));
 
     //!Last CacheChange Sequence Number added to the History.
     SequenceNumber_t m_lastCacheChangeSeqNum;
     //!Pointer to the associated RTPSWriter;
     RTPSWriter* mp_writer;
 };
 
-}
+} // namespace rtps
```
[the remark this answers is not in the document]
Comment (MiguelBarro) [reply include/fastdds/rtps/history/WriterHistory.h:129]: what difference does it make? 😉

File: src/cpp/rtps/builtin/BuiltinProtocols.cpp [hunk 7]
```diff
@@ -21,7 +21,8 @@
 #include <fastdds/rtps/common/Locator.h>
 
 #include <fastdds/rtps/builtin/discovery/participant/PDPSimple.h>
-#include "./discovery/participant/PDPClient2.hpp"
+// #include "./discovery/participant/PDPClient2.hpp"
+#include <fastdds/rtps/builtin/discovery/participant/PDPClient.h>
```
[the remark this answers is not in the document]
Comment (MiguelBarro) [reply src/cpp/rtps/builtin/BuiltinProtocols.cpp:25]: in the near future (less than a fortnight) all client functionality will move to PDPClient2.hpp

File: src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp [hunk 8]
```diff
@@ -354,6 +408,14 @@ void DiscoveryDataBase::create_writers_from_change(
...
                     }
                 }
+                pit = participants_.find(writer_guid.guidPrefix);
+                if (pit != participants_.end())
+                {
+                    if (!pit->second.is_matched(reader_it.guidPrefix))
+                    {
+                        pit->second.add_or_update_ack_participant(reader_it.guidPrefix);
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (MiguelBarro) [reply src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp:418]: not duplicated but we can save a search.

File: src/cpp/rtps/history/History.cpp (hunk 2 above)
Comment (MiguelBarro) [reply src/cpp/rtps/history/History.cpp:56]: this way I assure that lack of mutex leads to assertion on next iterator check (at least for Visual Studio and Debug config) 😎😎

File: src/cpp/rtps/history/History.cpp (hunk 3 above)
Comment (MiguelBarro) [reply src/cpp/rtps/history/History.cpp:220]: again?

File: src/cpp/rtps/builtin/discovery/participant/timedevent/DSClientEvent.cpp (hunk 1 above)
Comment (MiguelBarro) [reply src/cpp/rtps/builtin/discovery/participant/timedevent/DSClientEvent.cpp:87]: diversity is good 👨🏿

File: src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp]: Careful! It's not duplicated. One updates the reader's participant and the other updated the writer's participant

[the remark this answers is not in the document]
Comment (rsanchez15) [reply src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp]: You're right, I don't know why I saw it identical. And good catch @MiguelBarro removing that extra find.

**Review by rsanchez15 — APPROVED** [review]
LGTM

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2793/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/144/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2565/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2805/)

**Pull request merged**